Lib/unittest/test/testmock/testasync.py

```diff
@@ -378,35 +378,85 @@ def test_add_side_effect_iterable(self):
 class AsyncContextManagerTest(unittest.TestCase):
 
     class WithAsyncContextManager:
-
         async def __aenter__(self, *args, **kwargs):
             return self
 
         async def __aexit__(self, *args, **kwargs):
             pass
 
-    def test_magic_methods_are_async_mocks(self):
-        mock = MagicMock(self.WithAsyncContextManager())
-        self.assertIsInstance(mock.__aenter__, AsyncMock)
-        self.assertIsInstance(mock.__aexit__, AsyncMock)
+    class WithSyncContextManager:
+        def __enter__(self, *args, **kwargs):
+            return self
+
+        def __exit__(self, *args, **kwargs):
+            pass
+
+    class ProductionCode:
+        # Example real-world(ish) code
+        def __init__(self):
+            self.session = None
+
+        async def main(self):
+            async with self.session.post('https://python.org') as response:
+                val = await response.json()
+                return val
+
+    def test_async_magic_methods_are_async_mocks_with_magicmock(self):
+        cm_mock = MagicMock(self.WithAsyncContextManager)
+        self.assertIsInstance(cm_mock.__aenter__, AsyncMock)
+        self.assertIsInstance(cm_mock.__aexit__, AsyncMock)
+
+    def test_magicmock_has_async_magic_methods(self):
+        cm = MagicMock(name='magic_cm')
+        self.assertTrue(hasattr(cm, "__aenter__"))
+        self.assertTrue(hasattr(cm, "__aexit__"))
+
+    def test_magic_methods_are_async_functions(self):
+        cm = MagicMock(name='magic_cm')
+        self.assertTrue(asyncio.iscoroutinefunction(cm.__aenter__))
+        self.assertTrue(asyncio.iscoroutinefunction(cm.__aexit__))
+        # These should pass but cause warnings to be raised
+        # self.assertTrue(inspect.isawaitable(cm.__aenter__()))
```

Comment:
Can we just await on the method for the tests? Would prefer not having commented code.

```
+        for meth in [cm.__aenter__, cm.__aexit__]:
+            awaitable = meth()
+            self.assertTrue(inspect.isawaitable(awaitable))
+            await awaitable
```

Reply:
This will fail because we are trying to await outside of an async function :( 

I haven't looked at IsolatedAsyncioTestCase yet! Would that allow us to have an `async def test_blah` that would solve this?

Reply:
> I haven't looked at IsolatedAsyncioTestCase yet! Would that allow us to have an async def test_blah that would solve this?

I just documented it yesterday ;) yes, you can write `async def test_blah` and that would be collected by test runner. It's like a drop in replacement to `TestCase` and in addition also takes coroutines/async functions as test cases.

Reply:
Perfect! That would make testing AsyncMock more robust, I've found a good number of places I would like to test with `await` but haven't been able to. Let's make it another issue, I can take these commented lines out and they will eventually get added back in with the new `IsolatedAsyncioTestCase`.

Reply:
`IsolatedAsyncioTestCase` is not used by any CPython test yet (except self-testing) but I want to rewrite the most of asyncio tests with its usage eventually.